### PR TITLE
[chromecast-caf-receiver] Align to SDK and docs (only breaking changes)

### PR DESCRIPTION
**Pull request by marcoreni**
This PR supersedes #54403 , and is the first step in bringing this package up to date with the [Web Receiver Documentation](https://developers.google.cn/cast/docs/reference/web_receiver)

NOTES:
- Where in doubt (for example, removed methods) I double-checked on the SDK to make sure that they weren't undocumented methods.
- **Only** breaking changes are included in this PR. Following PRs will address new fields and documentation updates, but won't require a new major version.
- Seems like prettier was not run in a while so some of the changes are due to that.
- I followed [Google Javascript Styleguide](https://google.github.io/styleguide/javascriptguide.xml?showone=Comments#Comments) for comments (again, only for changed declarations). Following PRs will address docs cleanup and jsdoc syntax errors.

Please fill in this template.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [X] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [X] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header. - NOTE: I changed the major version since there are some breaking changes (removed methods, typos, errors). Since this is a CDN provided library, there is a versioning in place but it's not possible to pin a specific version in the clients; it's only possible to test the preview version. Reference: https://developers.google.cn/cast/preview-url , https://developers.google.cn/cast/docs/release-notes

### CODE REVIEW

File: types/chromecast-caf-receiver/cast.framework.d.ts [hunk 1]
```diff
@@ -559,19 +564,20 @@ export class PlayerManager {
     getDurationSec(): number;
 
     /**
-     * Returns live seekable range with start and end time in seconds. The values are media time based.
+     * @returns live seekable range with start and end time in seconds. The values
+     * are media time based.
```
Comment (Serabe) [inline types/chromecast-caf-receiver/cast.framework.d.ts:568]: Shouldn't this line be indented?

File: types/chromecast-caf-receiver/cast.framework.events.d.ts [hunk 2]
```diff
@@ -272,60 +272,63 @@ export class Id3Event extends Event {
...
+ * @see https://developers.google.com/cast/docs/reference/web_receiver/cast.framework.events.EmsgEvent
  */
 export class EmsgEvent extends Event {
     constructor(emsgData: any);
 
     /**
      * The time that the event ends (in presentation time). Undefined if using legacy Dash support.
      */
-    endTime: any;
+    endTime?: number | undefined;
```
Comment (Serabe) [inline types/chromecast-caf-receiver/cast.framework.events.d.ts:284]: Redundant. `?` means already `| undefined`.

File: types/chromecast-caf-receiver/cast.framework.events.d.ts [hunk 3]
```diff
@@ -272,60 +272,63 @@ export class Id3Event extends Event {
...
 
     /**
      * The duration of the event (in units of timescale). Undefined if using legacy Dash support.
      */
-    eventDuration: any;
+    eventDuration?: number | undefined;
```
Comment (Serabe) [inline types/chromecast-caf-receiver/cast.framework.events.d.ts:289]: Redundant. `?` means already `| undefined`.

File: types/chromecast-caf-receiver/cast.framework.events.d.ts [hunk 4]
```diff
@@ -272,60 +272,63 @@ export class Id3Event extends Event {
...
 
     /**
      * A field identifying this instance of the message. Undefined if using legacy Dash support.
      */
-    id: any;
+    id?: number | undefined;
```
Comment (Serabe) [inline types/chromecast-caf-receiver/cast.framework.events.d.ts:294]: Redundant. `?` means already `| undefined`.

File: types/chromecast-caf-receiver/cast.framework.events.d.ts [hunk 5]
```diff
@@ -272,60 +272,63 @@ export class Id3Event extends Event {
...
 
     /**
-     * Body of the message. Undefined if using legacy Dash support.
+     * Body of the message, which can contain ID3 metadata if signaled by the
+     * schemeIdUri value. The schemeIdUri "https://aomedia.org/emsg/ID3" and
+     * "https://developer.apple.com/streaming/emsg-id3" will always signal ID3 metadata.
      */
-    messageData: any;
+    messageData?: Uint8Array | undefined;
```
Comment (Serabe) [inline types/chromecast-caf-receiver/cast.framework.events.d.ts:301]: Redundant. `?` means already `| undefined`.

File: types/chromecast-caf-receiver/cast.framework.events.d.ts [hunk 6]
```diff
@@ -272,60 +272,63 @@ export class Id3Event extends Event {
...
 
     /**
      * The offset that the event starts; relative to the start of the segment this is contained in (in units of timescale). Undefined if using legacy Dash support.
      */
-    presentationTimeDelta: any;
+    presentationTimeDelta?: number | undefined;
```
Comment (Serabe) [inline types/chromecast-caf-receiver/cast.framework.events.d.ts:306]: Redundant. `?` means already `| undefined`.

File: types/chromecast-caf-receiver/cast.framework.events.d.ts [hunk 7]
```diff
@@ -272,60 +272,63 @@ export class Id3Event extends Event {
...
 
     /**
      * Identifies the message scheme. Undefined if using legacy Dash support.
      */
-    schemeIdUri: any;
+    schemeIdUri?: string | undefined;
```
Comment (Serabe) [inline types/chromecast-caf-receiver/cast.framework.events.d.ts:311]: Redundant. `?` means already `| undefined`.

File: types/chromecast-caf-receiver/cast.framework.events.d.ts [hunk 8]
```diff
@@ -272,60 +272,63 @@ export class Id3Event extends Event {
...
 
     /**
      * The segment data. This is only defined if using legacy Dash support.
      */
-    segmentData: any;
+    segmentData?: Uint8Array | undefined;
```
Comment (Serabe) [inline types/chromecast-caf-receiver/cast.framework.events.d.ts:316]: Redundant. `?` means already `| undefined`.

File: types/chromecast-caf-receiver/cast.framework.events.d.ts [hunk 9]
```diff
@@ -272,60 +272,63 @@ export class Id3Event extends Event {
...
      * The time that the event starts (in presentation time). Undefined if using legacy Dash support.
      */
-    startTime: any;
+    startTime?: number | undefined;
 
     /**
      * Provides the timescale; in ticks per second. Undefined if using legacy Dash support.
      */
-    timescale: any;
+    timescale?: number | undefined;
```
Comment (Serabe) [inline types/chromecast-caf-receiver/cast.framework.events.d.ts:326]: Redundant. `?` means already `| undefined`.

File: types/chromecast-caf-receiver/cast.framework.events.d.ts [hunk 10]
```diff
@@ -272,60 +272,63 @@ export class Id3Event extends Event {
...
      * The segment data. This is only defined if using legacy Dash support.
      */
-    segmentData: any;
+    segmentData?: Uint8Array | undefined;
 
     /**
      * The time that the event starts (in presentation time). Undefined if using legacy Dash support.
      */
-    startTime: any;
+    startTime?: number | undefined;
```
Comment (Serabe) [inline types/chromecast-caf-receiver/cast.framework.events.d.ts:321]: Redundant. `?` means already `| undefined`.

File: types/chromecast-caf-receiver/cast.framework.events.d.ts [hunk 11]
```diff
@@ -423,13 +431,14 @@ export class BreaksEvent extends Event {
...
- * Event data for @see {@link EventType.BITRATE_CHANGED} event.
+ * Event data for {@link EventType.BITRATE_CHANGED} event.
+ * @see https://developers.google.com/cast/docs/reference/web_receiver/cast.framework.events.BitrateChangedEvent
  */
 export class BitrateChangedEvent extends Event {
     constructor(totalBitrate?: number);
 
     /** The bitrate of the media (audio and video) in bits per second. */
-    totalBitrate: number;
+    totalBitrate?: number | undefined;
```
Comment (Serabe) [inline types/chromecast-caf-receiver/cast.framework.events.d.ts:441]: Redundant. `?` means already `| undefined`.

File: types/chromecast-caf-receiver/cast.framework.messages.d.ts [hunk 12]
```diff
@@ -1460,7 +1484,7 @@ export class MusicTrackMediaMetadata {
      * Content images. Examples would include cover art or a thumbnail of the
      * currently playing media.
      */
-    images: Image[];
+    images?: Image[] | undefined;
```
Comment (Serabe) [inline types/chromecast-caf-receiver/cast.framework.messages.d.ts:1487]: Redundant. `?` means already `| undefined`.

File: types/chromecast-caf-receiver/cast.framework.messages.d.ts [hunk 13]
```diff
@@ -1797,7 +1826,7 @@ export class LoadRequestData extends RequestData {
     /**
      * Queue data.
      */
-    queueData: QueueData;
+    queueData?: QueueData | undefined;
```
Comment (Serabe) [inline types/chromecast-caf-receiver/cast.framework.messages.d.ts:1829]: Redundant. `?` means already `| undefined`.

File: types/chromecast-caf-receiver/cast.framework.messages.d.ts [hunk 14]
```diff
@@ -1985,11 +2016,14 @@ export class FetchItemsRequestData extends RequestData {
...
 /**
  * Extended media status information
+ * @see https://developers.google.com/cast/docs/reference/caf_receiver/cast.framework.messages.ExtendedMediaStatus
  */
 export class ExtendedMediaStatus {
-    constructor(playerState: MediaInformation, opt_media?: MediaInformation);
+    constructor(playerState: MediaInformation, opt_media?: MediaInformation, opt_mediaSessionId?: number);
 
-    media: MediaInformation;
+    media?: MediaInformation | undefined;
```
Comment (Serabe) [inline types/chromecast-caf-receiver/cast.framework.messages.d.ts:2024]: Redundant. `?` means already `| undefined`.

File: types/chromecast-caf-receiver/cast.framework.messages.d.ts [hunk 15]
```diff
@@ -1985,11 +2016,14 @@ export class FetchItemsRequestData extends RequestData {
...
+
+    mediaSessionId?: number | undefined;
```
Comment (Serabe) [inline types/chromecast-caf-receiver/cast.framework.messages.d.ts:2026]: Redundant. `?` means already `| undefined`.

File: types/chromecast-caf-receiver/cast.framework.messages.d.ts [hunk 16]
```diff
@@ -2140,35 +2174,40 @@ export class CustomCommandRequestData extends RequestData {
...
+ */
 export class BreakStatus {
-    constructor(currentBreakTime: number, currentBreakClipTime: number);
+    constructor(currentBreakTime: number | undefined, currentBreakClipTime: number | undefined);
 
     /**
      * Id of current break clip.
      */
-    breakClipId: string;
+    breakClipId?: string | undefined;
```
Comment (Serabe) [inline types/chromecast-caf-receiver/cast.framework.messages.d.ts:2188]: Redundant. `?` means already `| undefined`.

File: types/chromecast-caf-receiver/cast.framework.messages.d.ts [hunk 17]
```diff
@@ -2140,35 +2174,40 @@ export class CustomCommandRequestData extends RequestData {
...
 
     /**
      * Id of current break.
      */
-    breakId: string;
+    breakId?: string | undefined;
```
Comment (Serabe) [inline types/chromecast-caf-receiver/cast.framework.messages.d.ts:2193]: Redundant. `?` means already `| undefined`.

File: types/chromecast-caf-receiver/cast.framework.messages.d.ts [hunk 18]
```diff
@@ -2140,35 +2174,40 @@ export class CustomCommandRequestData extends RequestData {
...
 export class CloudMediaStatus extends MediaStatus {}
 
+/**
+ * Represents current status of break.
+ *
+ * @see https://developers.google.com/cast/docs/reference/caf_receiver/cast.framework.messages.BreakStatus
+ */
 export class BreakStatus {
-    constructor(currentBreakTime: number, currentBreakClipTime: number);
+    constructor(currentBreakTime: number | undefined, currentBreakClipTime: number | undefined);
```
Comment (Serabe) [inline types/chromecast-caf-receiver/cast.framework.messages.d.ts:2183]: Add `?` to both parameters and remove `| undefined`. That way we can omit the arguments if they are undefined while we can still pass undefined to them.

See this playground

https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABKMAKAhgfgFyLCAWwCMBTAJ0QB9FwATE4GME2gSkQG8AoRRMkqCDJJ0Abi4BfLilStxMgKxzp4VHQZMWrLisiwEyEBhx5CpMu269+g4YjGSVs+UaUu1Yeo2ZsdKaPBIwGjouPjE5FQ0nho+ljx8AkIi4lLSaMrBqG7pHl6abKJAA

File: types/chromecast-caf-receiver/cast.framework.messages.d.ts [hunk 19]
```diff
@@ -2140,35 +2174,40 @@ export class CustomCommandRequestData extends RequestData {
...
      * Id of current break.
      */
-    breakId: string;
+    breakId?: string | undefined;
 
     /**
      * Time in sec elapsed after current break clip starts.
      */
-    currentBreakClipTime: number;
+    currentBreakClipTime?: number | undefined;
```
Comment (Serabe) [inline types/chromecast-caf-receiver/cast.framework.messages.d.ts:2198]: Redundant.

File: types/chromecast-caf-receiver/cast.framework.messages.d.ts [hunk 20]
```diff
@@ -2140,35 +2174,40 @@ export class CustomCommandRequestData extends RequestData {
...
 
     /**
      * Time in sec elapsed after current break starts.
      */
-    currentBreakTime: number;
+    currentBreakTime?: number | undefined;
```
Comment (Serabe) [inline types/chromecast-caf-receiver/cast.framework.messages.d.ts:2203]: Redundant.

File: types/chromecast-caf-receiver/cast.framework.messages.d.ts [hunk 21]
```diff
@@ -2140,35 +2174,40 @@ export class CustomCommandRequestData extends RequestData {
...
 
     /**
      * The time in sec when this break clip becomes skippable.
      * 5 means that end user can skip this break clip after 5 seconds.
      * If this field is not defined; it means that current break clip is not skippable.
      */
-    whenSkippable: number;
+    whenSkippable?: number | undefined;
```
Comment (Serabe) [inline types/chromecast-caf-receiver/cast.framework.messages.d.ts:2210]: Redundant.

File: types/chromecast-caf-receiver/cast.framework.messages.d.ts [hunk 22]
```diff
@@ -2264,3 +2303,73 @@ export class Break {
...
+/**
+ * An audiobook chapter description.
+ * @see https://developers.google.com/cast/docs/reference/caf_receiver/cast.framework.messages.AudiobookChapterMediaMetadata
+ */
+export class AudiobookChapterMediaMetadata {
+    constructor();
+    /**
+     * Audiobook title.
+     */
+    bookTitle?: string | undefined;
```
Comment (Serabe) [inline types/chromecast-caf-receiver/cast.framework.messages.d.ts:2316]: Redundant.

File: types/chromecast-caf-receiver/cast.framework.messages.d.ts [hunk 23]
```diff
@@ -2264,3 +2303,73 @@ export class Break {
...
+
+    /**
+     * Chapter number, used for display purposes.
+     */
+    chapterNumber?: number | undefined;
```
Comment (Serabe) [inline types/chromecast-caf-receiver/cast.framework.messages.d.ts:2321]: Redundant.

File: types/chromecast-caf-receiver/cast.framework.messages.d.ts [hunk 24]
```diff
@@ -2264,3 +2303,73 @@ export class Break {
...
+
+    /**
+     * Chapter title.
+     */
+    chapterTitle?: string | undefined;
```
Comment (Serabe) [inline types/chromecast-caf-receiver/cast.framework.messages.d.ts:2326]: Redundant.

File: types/chromecast-caf-receiver/cast.framework.messages.d.ts [hunk 25]
```diff
@@ -2264,3 +2303,73 @@ export class Break {
...
+    /**
+     * Content rating.
+     * Note as of July 2021: docs do not specify a type
+     */
+    contentRating?: any;
+
+    /**
+     * Chapter or book cover art.
+     */
+    images?: Image[] | undefined;
```
Comment (Serabe) [inline types/chromecast-caf-receiver/cast.framework.messages.d.ts:2337]: Redundant.

File: types/chromecast-caf-receiver/cast.framework.messages.d.ts [hunk 26]
```diff
@@ -2264,3 +2303,73 @@ export class Break {
...
+
+    /**
+     * Audiobook title, for backward compatibility.
+     */
+    subtitle?: string | undefined;
```
Comment (Serabe) [inline types/chromecast-caf-receiver/cast.framework.messages.d.ts:2342]: Redundant.

File: types/chromecast-caf-receiver/cast.framework.messages.d.ts [hunk 27]
```diff
@@ -2264,3 +2303,73 @@ export class Break {
...
+
+    /**
+     * Chapter title, for backward compatibility.
+     */
+    title?: string | undefined;
```
Comment (Serabe) [inline types/chromecast-caf-receiver/cast.framework.messages.d.ts:2347]: Redundant.

File: types/chromecast-caf-receiver/cast.framework.messages.d.ts [hunk 28]
```diff
@@ -2264,3 +2303,73 @@ export class Break {
...
+/**
+ * An audiobook container description.
+ * @see https://developers.google.com/cast/docs/reference/caf_receiver/cast.framework.messages.AudiobookContainerMetadata
+ */
+export class AudiobookContainerMetadata {
+    constructor();
+    /**
+     * Book authors.
+     */
+    authors?: string[] | undefined;
```
Comment (Serabe) [inline types/chromecast-caf-receiver/cast.framework.messages.d.ts:2359]: Redundant.

File: types/chromecast-caf-receiver/cast.framework.messages.d.ts [hunk 29]
```diff
@@ -2264,3 +2303,73 @@ export class Break {
...
+
+    /**
+     * Book narrators.
+     */
+    narrators?: string[] | undefined;
```
Comment (Serabe) [inline types/chromecast-caf-receiver/cast.framework.messages.d.ts:2364]: Redundant.

File: types/chromecast-caf-receiver/cast.framework.messages.d.ts [hunk 30]
```diff
@@ -2264,3 +2303,73 @@ export class Break {
...
+
+    /**
+     * Book publisher.
+     */
+    publisher?: string | undefined;
```
Comment (Serabe) [inline types/chromecast-caf-receiver/cast.framework.messages.d.ts:2369]: Redundant.

File: types/chromecast-caf-receiver/cast.framework.messages.d.ts [hunk 31]
```diff
@@ -2264,3 +2303,73 @@ export class Break {
...
+
+    /**
+     * ISO 8601 date when the book was released; e.g. 2014-02-10.
+     */
+    releaseDate?: string | undefined;
```
Comment (Serabe) [inline types/chromecast-caf-receiver/cast.framework.messages.d.ts:2374]: Redundant.

File: types/chromecast-caf-receiver/cast.framework.system.d.ts [hunk 32]
```diff
@@ -80,6 +80,40 @@ export enum DisconnectReason {
...
+    DIAL = 'DIAL',
+    /**
+     * App was launched by Cast V2 request.
+     */
+    CAST = 'CAST',
+    /**
+     * App was launched by assistant request (e.g. voice command).
+     */
+    CLOUD = 'CLOUD',
+}
```
Comment (Serabe) [inline types/chromecast-caf-receiver/cast.framework.system.d.ts:104]: Alphabetical order, please.

File: types/chromecast-caf-receiver/cast.framework.system.d.ts [hunk 33]
```diff
@@ -80,6 +80,40 @@ export enum DisconnectReason {
...
+
+/**
+ * Represents the current visibility state reported by the platform.
+ * It may be UNKNOWN if the cast platform was unable to determine the state yet.
+ * @see https://developers.google.com/cast/docs/reference/web_receiver/cast.framework.system#.VisibilityState
+ */
+export enum VisibilityState {
+    VISIBLE = 'visible',
+    NOT_VISIBLE = 'notvisible',
+    UNKNOWN = 'unknown',
```
Comment (Serabe) [inline types/chromecast-caf-receiver/cast.framework.system.d.ts:114]: Alphabetical, please.

**Review by Serabe — CHANGES_REQUESTED** [review]
Follow up about `field?: type | undefined` in a comment.

File: types/chromecast-caf-receiver/cast.framework.events.d.ts (hunk 2 above)
Comment (marcoreni) [reply types/chromecast-caf-receiver/cast.framework.events.d.ts:284]: Please see https://github.com/DefinitelyTyped/DefinitelyTyped/pull/54357

File: types/chromecast-caf-receiver/cast.framework.events.d.ts (hunk 3 above)
Comment (marcoreni) [reply types/chromecast-caf-receiver/cast.framework.events.d.ts:289]: Please see https://github.com/DefinitelyTyped/DefinitelyTyped/pull/54357

**Comment by marcoreni** [comment]
`{type} | undefined` is part of changes added to DefinitelyTyped in preparation of Typescript 4.4: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/54357 - https://github.com/microsoft/dtslint/pull/335

**Comment by Serabe** [comment]
In TS optional fields are fields that might not be there for an object to be considered of a class. [Example playground](https://www.typescriptlang.org/play?#code/MYGwhgzhAECC0G8BQ1oDMCWBTEATA-AFzQQAuAThgHYDmA3EitMAPZVnkCuwpL5AFGmIdqNAJSImqUgAsMEAHSYcuaAF50DVAF8kupK3aloYYvA0JtDQxBYgsCkCxr8wS7HjFA)

You can also check that the autocomplete, completes `the info to type `field?: string | undefined`.

2021-07-21 at 16 26

OTOH, if we declared a field as `field: type | undefined`, the presence of that field is required for an object to be considered of that class. [See this other playground](https://www.typescriptlang.org/play?#code/MYGwhgzhAECC0G8BQ1oDMCWBTEATAXNBAC4BOGAdgObQA+0ArhblphVrgNxIrTAD2FEqQbBi-UgAo0hYZSoBKRL1TEAFhggA6TDlzQAvOm6oAvknNIBQ4tDCF4RhKe7WI-EFi0h+VSWB1sPAUgA)

2021-07-21 at 16 28

I think that, in our scenario, the best approach is optional field because:

1. A non-existent field is always accessible in JavaScript (and therefore in TS). I mean, `const a = {}; a.whatever` does not raise any error, just returns `undefined`.
2. Having less conditions for an object to be considered of a class is better in this situation, making it easier for people to test bits of the application outside the receiver itself.

So my vote if for all the `field?: type | undefined` to be just `field?: type`, unless anyone wants to force the fields for whatever reason I cannot think of right now.

**Comment by Serabe** [comment]
I totally missed that announcement. Thank you! Can you fix the rest so I can approve it?

File: types/chromecast-caf-receiver/cast.framework.d.ts (hunk 1 above)
Comment (marcoreni) [reply types/chromecast-caf-receiver/cast.framework.d.ts]: Correct! Done.

File: types/chromecast-caf-receiver/cast.framework.messages.d.ts (hunk 18 above)
Comment (marcoreni) [reply types/chromecast-caf-receiver/cast.framework.messages.d.ts]: I checked the caf_receiver sdk and in fact  `new cast.framework.messages.BreakStatus()` works. I've updated the declaration (I've kept `| undefined` as per TS4.4)

File: types/chromecast-caf-receiver/cast.framework.system.d.ts (hunk 32 above)
Comment (marcoreni) [reply types/chromecast-caf-receiver/cast.framework.system.d.ts]: Done.

File: types/chromecast-caf-receiver/cast.framework.system.d.ts (hunk 33 above)
Comment (marcoreni) [reply types/chromecast-caf-receiver/cast.framework.system.d.ts]: Done.

**Review by Serabe — APPROVED** [review]

**Comment by Serabe** [comment]
Sorry for the delay. My personal life is chaotic right now.

**Comment by marcoreni** [comment]
Ready to merge

**Comment by wory04** [comment]
> Ready to merge

@marcoreni Thanks! Are you planning to put up another PR with what's left from the [previous one](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/54403)? :)

**Comment by marcoreni** [comment]
@wory04 yes, the idea is to create some more PRs adding missing fields, fixing docs and so on. I'll keep 'em smaller than this one so that they'll be easier to review and merge :)

**Pull request merged**